### PR TITLE
Remove S.Linq dependency from S.R.Metadata

### PR DESCRIPTION
**Pull request by nguerrera**
@tmat @davidwrighton

### CODE REVIEW

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EnumerableExtensions.cs [hunk 1]
```diff
@@ -3,40 +3,48 @@
...
+    /// </summary>
     internal static class EnumerableExtensions
     {
-        private class ComparisonComparer<T> : Comparer<T>
+        public static T FirstOrDefault<T>(this ImmutableArray<T> collection, Func<T, bool> predicate)
         {
-            private readonly Comparison<T> _compare;
-
-            public ComparisonComparer(Comparison<T> compare)
+            foreach (var item in collection)
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EnumerableExtensions.cs]: Nit: var => T

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EnumerableExtensions.cs [hunk 2]
```diff
@@ -3,40 +3,48 @@
...
+            {
+                yield return selector(item);
+            }
         }
 
-        public static IOrderedEnumerable<T> OrderBy<T>(this IEnumerable<T> source, IComparer<T> comparer)
+        public static T Last<T>(this ImmutableArray<T>.Builder source)
         {
-            return source.OrderBy(Functions<T>.Identity, comparer);
+            return source[source.Count - 1];
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EnumerableExtensions.cs]: Do we care what exception is thrown here when the builder is empty?

**Comment by nguerrera** [comment]
@stephentoub Thanks for reviewing. 

> Does that mean it was being pulled in as a dependency implicitly?

Yes, it's coming via at least System.Collections.Immutable AFAICT, but we have #14369 tracking moving ImmutableArray (all that is used by S.R.M) down.

I played with this some more by attempting a build against System.Private.CoreLib plus shared source files. It turns out there's a bigger tail here (Compression, Dictionary, BitConverter, BinaryReader, StrongBox, and HashSet are among non corelib dependencies). https://github.com/dotnet/corefx/commit/0f721870c22e649db2573355f624b958715ddce4 is my work in progress to shake out the full list of things that we use that are not in corelib. It is still incomplete. 

Linq was just the most surprising thing to show up in assembly-level dependency analysis, but this proves that's trivial. I'm going to see how the discussion shakes out on #14369 before pursuing any actual product changes. So I'll close this for now and only bother with the cleanup vs these review notes once we're sure we want to push forward on getting this lib to build without the non-corelib dependencies.

FYI, @davidwrighton

**Comment by tmat** [comment]
- [ ] BinaryReader: this should get rid of it: https://github.com/dotnet/corefx/issues/7996
- [ ] StrongBox is trivial
- [ ] BitConverter - I think we only use it to display byte[] in debugger display, should be easy to avoid
- [ ] HashSet, Dictionary - huh? nothing in System.Private.CoreLib uses hashing???
- [ ] Compression - need for embedded Portable PDBs

**Pull request closed without merging**